File: python/src/contriwork_market_data/_rate_limit.py
"""Per-adapter token-bucket rate limiter + retry — CONTRACT.md §6.

The orchestrator owns one :class:`TokenBucket` per adapter (capacity =
``Capability.rate_limit_per_minute``). Operations queue on the bucket and
retry with exponential backoff when an adapter raises
:class:`~contriwork_market_data.errors.RateLimitedError`.
"""

from __future__ import annotations

import secrets
from collections.abc import Awaitable, Callable
from typing import TypeVar

from ._clock import Clock
from .config import RateLimitConfig
from .errors import RateLimitedError

__all__ = ["TokenBucket", "run_with_retry"]

T = TypeVar("T")


class TokenBucket:
    """Classic refill token bucket; not thread-safe."""

    def __init__(self, *, capacity: int, refill_per_second: float, clock: Clock) -> None:
        if capacity < 1:
            raise ValueError("capacity must be >= 1")
        if refill_per_second < 0:
            raise ValueError("refill_per_second must be >= 0")
        self._capacity = capacity
        self._refill = refill_per_second
        self._tokens = float(capacity)
        self._last = clock.monotonic()
        self._clock = clock

    def _refill_tokens(self) -> None:
        now = self._clock.monotonic()
        elapsed = max(0.0, now - self._last)
        self._tokens = min(float(self._capacity), self._tokens + elapsed * self._refill)
        self._last = now

    def try_acquire(self, tokens: int = 1) -> bool:
        self._refill_tokens()
        if self._tokens + 1e-9 >= tokens:
            self._tokens -= tokens
            return True
        return False

    def time_until_available(self, tokens: int = 1) -> float:
        self._refill_tokens()
        deficit = tokens - self._tokens
        if deficit <= 0:
            return 0.0
        if self._refill <= 0:
            # No refill: the bucket never recovers; treat as infinite wait.
            return float("inf")
        return deficit / self._refill


async def run_with_retry[T](
    fn: Callable[[], Awaitable[T]],
    *,
    config: RateLimitConfig,
    clock: Clock,
    bucket: TokenBucket | None = None,
) -> T:
    """Invoke ``fn`` with rate-limit-aware retry.

    Behavior:
    - If ``bucket`` is provided, wait for a token before each attempt.
    - On :class:`RateLimitedError`, sleep with jittered exponential backoff
      and retry, up to ``config.max_retry_attempts`` extra attempts.
    - Other exceptions propagate immediately.
    """
    attempts = 0
    backoff_s = max(0.0, config.initial_backoff_s)
    while True:
        if bucket is not None:
            wait_s = bucket.time_until_available()
            if wait_s > 0:
                # Inf would block forever; cap defensively at max_backoff_s.
                if wait_s == float("inf"):
                    wait_s = config.max_backoff_s
                await clock.sleep(wait_s)
            bucket.try_acquire()
        try:
            return await fn()
        except RateLimitedError:
            if attempts >= config.max_retry_attempts:
                raise
            attempts += 1
            sleep_s = min(backoff_s, config.max_backoff_s)
            if config.jitter:
                # ``secrets.SystemRandom`` is overkill for jitter but the
                # plain ``random`` module is flagged by bandit (S311) for
                # crypto use. ``SystemRandom`` keeps lint clean and the cost
                # is irrelevant compared to network latency.
                sleep_s *= 0.5 + secrets.SystemRandom().random()
            await clock.sleep(sleep_s)
            backoff_s = min(config.max_backoff_s, max(backoff_s, 0.001) * 2.0)

File: typescript/src/internal/rate-limit.ts
/**
 * Per-adapter token-bucket rate limiter + retry runner — CONTRACT.md §6.
 */
import type { Clock } from "../clock.js";
import type { RateLimitConfig } from "../config.js";
import { RateLimitedError } from "../errors.js";

/** Classic refill token bucket; not concurrency-safe. */
export class TokenBucket {
  private readonly capacity: number;
  private readonly refillPerSecond: number;
  private readonly clock: Clock;
  private tokens: number;
  private last: number;

  public constructor(capacity: number, refillPerSecond: number, clock: Clock) {
    if (capacity < 1) {
      throw new Error("capacity must be >= 1");
    }
    if (refillPerSecond < 0) {
      throw new Error("refillPerSecond must be >= 0");
    }
    this.capacity = capacity;
    this.refillPerSecond = refillPerSecond;
    this.clock = clock;
    this.tokens = capacity;
    this.last = clock.monotonic();
  }

  /** Take a token if available. */
  public tryAcquire(): boolean {
    this.refill();
    if (this.tokens + 1e-9 >= 1) {
      this.tokens -= 1;
      return true;
    }
    return false;
  }

  /** Seconds until a token is available (`Infinity` if it never refills). */
  public timeUntilAvailable(): number {
    this.refill();
    const deficit = 1 - this.tokens;
    if (deficit <= 0) {
      return 0;
    }
    return this.refillPerSecond <= 0
      ? Infinity
      : deficit / this.refillPerSecond;
  }

  private refill(): void {
    const now = this.clock.monotonic();
    const elapsed = Math.max(0, now - this.last);
    this.tokens = Math.min(
      this.capacity,
      this.tokens + elapsed * this.refillPerSecond,
    );
    this.last = now;
  }
}

/**
 * Invoke `operation` with rate-limit-aware retry. On `RateLimitedError`, sleep
 * with jittered exponential backoff and retry up to `maxRetryAttempts` times.
 *
 * @param operation - The async operation to run.
 * @param config - Rate-limit configuration.
 * @param clock - Time source.
 * @param bucket - Optional token bucket to queue on.
 * @param signal - Cancellation signal.
 * @returns The operation result.
 */
export async function runWithRetry<TResult>(
  operation: () => Promise<TResult>,
  config: RateLimitConfig,
  clock: Clock,
  bucket: TokenBucket | undefined,
  signal?: AbortSignal,
): Promise<TResult> {
  let attempts = 0;
  let backoff = Math.max(0, config.initialBackoffSeconds);
  for (;;) {
    if (bucket !== undefined) {
      let wait = bucket.timeUntilAvailable();
      if (wait > 0) {
        if (wait === Infinity) {
          wait = config.maxBackoffSeconds;
        }
        await clock.sleep(wait, signal);
      }
      bucket.tryAcquire();
    }

    try {
      return await operation();
    } catch (err) {
      if (
        !(err instanceof RateLimitedError) ||
        attempts >= config.maxRetryAttempts
      ) {
        throw err;
      }
      attempts += 1;
      let sleepSeconds = Math.min(backoff, config.maxBackoffSeconds);
      if (config.jitter) {
        sleepSeconds *= 0.5 + Math.random();
      }
      await clock.sleep(sleepSeconds, signal);
      backoff = Math.min(
        config.maxBackoffSeconds,
        Math.max(backoff, 0.001) * 2,
      );
    }
  }
}
